### PR TITLE
PERF: drop GIL during long-running proj database calls

### PR DESCRIPTION
**Pull request by tpwrules**
These calls can take hundreds of milliseconds to access the PROJ databases, during which time no other Python code in any other thread can run. Use `nogil` to avoid this effect.

This should be safe for these calls as they all take a separate PROJ threading context.

It's quite probable there are other calls which could use this, but these negatively affected my application and I was able to identify them. I would welcome additions to the set of calls to wrap.

 - [ ] Closes #xxxx
 - [ ] Tests added
 - [ ] Fully documented, including `history.rst` for all changes and `api/*.rst` for new API

### CODE REVIEW

File: pyproj/_transformer.pyx
Comment (djhoese) [inline pyproj/_transformer.pyx:335]: Just for my own edification, does the declaration need to be moved outside the `with nogil` because Cython doesn't like it?

**Review by djhoese — APPROVED** [review]
I don't know enough about the thread-safety of PROJ, but I like the idea of this optimization. @tpwrules do you have any profiling results (pretty plots?) that show the execution time improvements you've seen by doing this?

Of course, something this low-level should really get approval from @snowman2 before merging.

File: pyproj/_transformer.pyx
Comment (tpwrules) [reply pyproj/_transformer.pyx:335]: Precisely. cdefs are only allowed directly within a function and not within a `with` or `try` block.

**Comment by tpwrules** [comment]
It doesn't improve execution time, just makes other threads in the process more stable. It's a really hard thing to benchmark and measure unfortunately. I just was able to observe the effect on my program, but I can't really share it here.

**Comment by jjimenezshaw** [comment]
> I don't know enough about the thread-safety of PROJ

PROJ does not deal with thread safety internally. It should be managed by the user via the "context". The same context must not be used simultaneously by two different threads.
(PROJ uses the "context" to keep there caches and any global initialization. So new context, new initialization. The access to the sqlite db -proj.db- is in read only mode, and managed by sqlite3 to avoid collisions)

File: pyproj/database.pyx
Comment (snowman2) [inline pyproj/database.pyx:417]: I am surprised cython didn't complain about using `bool` with `nogil`.

**Comment by snowman2** [comment]
Thanks @tpwrules 👍

**Comment by snowman2** [comment]
I would be interested to see if doing this with `proj_create` calls helps as well.

File: pyproj/database.pyx
Comment (tpwrules) [reply pyproj/database.pyx:417]: I'm assuming it's valid if it did not complain, but I'm not 100% sure.

**Pull request merged**
